cgen error: assign the spawn or go expr to the []thread

### Describe the bug

```
v run a.v
==================
/tmp/v_501/a.12000158363956472078.tmp.c:2138:8: error: expected ';' at end of declaration
        int a1 });
              ^
              ;
/tmp/v_501/a.12000158363956472078.tmp.c:2138:10: error: expected identifier or '('
        int a1 });
                ^
/tmp/v_501/a.12000158363956472078.tmp.c:2139:20: error: expected ')'
         _v_dump_expr_int(_SLIT("a.v"), 6, _SLIT("a"), a);
                          ^
/tmp/v_501/a.12000158363956472078.tmp.c:183:27: note: expanded from macro '_SLIT'
#define _SLIT(s) ((string){.str=(byteptr)("" s), .len=(sizeof(s)-1), .is_lit=1})
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

### Reproduction Steps

```v
fn main() {
	mut threads := []thread{len: 1}

	threads[0] = spawn fn () {
		a := 1
		dump(a)
		assert true
	}()

	threads.wait()
}
```

### Expected Behavior

no error.

### Current Behavior

cgen error.

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.4.1 09006fa

### Environment details (OS name and version, etc.)

V full version: V 0.4.1 09006fa
OS: macos, macOS, 13.5.2, 22G91
Processor: 8 cpus, 64bit, little endian, Intel(R) Core(TM) i7-8569U CPU @ 2.80GHz

getwd: /usr/local/v
vexe: /usr/local/v/v
vexe mtime: 2023-09-16 08:46:34

vroot: OK, value: /usr/local/v
VMODULES: OK, value: /Users/shove/.vmodules
VTMP: OK, value: /tmp/v_501

Git version: git version 2.39.2 (Apple Git-143)
Git vroot status: weekly.2023.30-315-g09006fa5
.git/config present: true

CC version: Apple clang version 14.0.3 (clang-1403.0.22.14.1)
thirdparty/tcc status: thirdparty-macos-amd64 46662e20
> [!IMPORTANT]
> You can vote for this issue using the 👍 reaction. More votes increase the issue's priority
> for developers.
>
> Take into account that only the 👍 reaction counts as a vote.
> Only reactions to the issue itself will be counted as votes, not comments.

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -3269,16 +3269,22 @@ fn (mut g Gen) expr(node_ ast.Expr) {
 			}
 		}
 		ast.SpawnExpr {
+			old_is_arraymap_set := g.is_arraymap_set
+			g.is_arraymap_set = false
 			g.spawn_and_go_expr(node, .spawn_)
+			g.is_arraymap_set = old_is_arraymap_set
 		}
 		ast.GoExpr {
 			// XTODO this results in a cgen bug, order of fields is broken
 			// g.spawn_and_go_expr(ast.SpawnExpr{node.pos, node.call_expr, node.is_expr},
+			old_is_arraymap_set := g.is_arraymap_set
+			g.is_arraymap_set = false
 			g.spawn_and_go_expr(ast.SpawnExpr{
 				pos: node.pos
 				call_expr: node.call_expr
 				is_expr: node.is_expr
 			}, .go_)
+			g.is_arraymap_set = old_is_arraymap_set
 		}
 		ast.Ident {
 			g.ident(node)

--- a/vlib/v/tests/thread_array_test.v
+++ b/vlib/v/tests/thread_array_test.v
@@ -1,0 +1,11 @@
+fn test_assign_spawn_to_element() {
+	mut threads := []thread{len: 1}
+
+	threads[0] = spawn fn () {
+		a := 1
+		dump(a)
+		assert true
+	}()
+
+	threads.wait()
+}

